Gradle autocompletion (ctrl + space) fails with error

Extension Name: vscode-gradle
Extension Version: v3.16.2024080611 (pre-release)
OS Version: Windows 11
VSCode version: nightly

**Describe the bug**
The extension or language server is failing completely when trying to do anything in a build.gradle file.
Trying to work on a minecraft mod with neoforge ModDevGradle template from below.

**To Reproduce**
`git clone https://github.com/NeoForgeMDKs/MDK-1.21-ModDevGradle.git`
go into build.gradle and press ctrl + space to autocomplete anywhere

**Expected behavior**
should show autocompletion list

**Screenshots**
![image](https://github.com/user-attachments/assets/7de55bfa-6fbc-4d0c-bb25-e79f3fecd59b)


**Output from "Gradle for Java"**
```
[Error - 2:06:37 PM] Request textDocument/completion failed.
  Message: Internal error.
  Code: -32603 
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 11 more
Caused by: java.lang.NullPointerException: Cannot invoke "String.equals(Object)" because the return value of "org.codehaus.groovy.ast.expr.MethodCallExpression.getMethodAsString()" is null
	at com.microsoft.gradle.compile.CompletionVisitor.visitMethodCallExpression(CompletionVisitor.java:107)
	at org.codehaus.groovy.ast.expr.MethodCallExpression.visit(MethodCallExpression.java:78)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitExpressionStatement(ClassCodeVisitorSupport.java:244)
	at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:41)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitBlockStatement(CodeVisitorSupport.java:86)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitBlockStatement(ClassCodeVisitorSupport.java:208)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitClosureExpression(CodeVisitorSupport.java:239)
	at org.codehaus.groovy.ast.expr.ClosureExpression.visit(ClosureExpression.java:110)
	at org.codehaus.groovy.ast.GroovyCodeVisitor.visitListOfExpressions(GroovyCodeVisitor.java:215)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitTupleExpression(CodeVisitorSupport.java:249)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitArgumentlistExpression(CodeVisitorSupport.java:367)
	at org.codehaus.groovy.ast.expr.ArgumentListExpression.visit(ArgumentListExpression.java:75)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitMethodCallExpression(CodeVisitorSupport.java:186)
	at com.microsoft.gradle.compile.CompletionVisitor.visitMethodCallExpression(CompletionVisitor.java:120)
	at org.codehaus.groovy.ast.expr.MethodCallExpression.visit(MethodCallExpression.java:78)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitExpressionStatement(ClassCodeVisitorSupport.java:244)
	at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:41)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitBlockStatement(CodeVisitorSupport.java:86)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitBlockStatement(ClassCodeVisitorSupport.java:208)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitClosureExpression(CodeVisitorSupport.java:239)
	at org.codehaus.groovy.ast.expr.ClosureExpression.visit(ClosureExpression.java:110)
	at org.codehaus.groovy.ast.GroovyCodeVisitor.visitListOfExpressions(GroovyCodeVisitor.java:215)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitTupleExpression(CodeVisitorSupport.java:249)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitArgumentlistExpression(CodeVisitorSupport.java:367)
	at org.codehaus.groovy.ast.expr.ArgumentListExpression.visit(ArgumentListExpression.java:75)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitMethodCallExpression(CodeVisitorSupport.java:186)
	at com.microsoft.gradle.compile.CompletionVisitor.visitMethodCallExpression(CompletionVisitor.java:120)
	at org.codehaus.groovy.ast.expr.MethodCallExpression.visit(MethodCallExpression.java:78)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitExpressionStatement(CodeVisitorSupport.java:117)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitExpressionStatement(ClassCodeVisitorSupport.java:244)
	at org.codehaus.groovy.ast.stmt.ExpressionStatement.visit(ExpressionStatement.java:41)
	at org.codehaus.groovy.ast.CodeVisitorSupport.visitBlockStatement(CodeVisitorSupport.java:86)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitBlockStatement(ClassCodeVisitorSupport.java:208)
	at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:72)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitClassCodeContainer(ClassCodeVisitorSupport.java:157)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitConstructorOrMethod(ClassCodeVisitorSupport.java:130)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitMethod(ClassCodeVisitorSupport.java:125)
	at org.codehaus.groovy.ast.ClassNode.visitMethods(ClassNode.java:1190)
	at org.codehaus.groovy.ast.ClassNode.visitContents(ClassNode.java:1183)
	at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitClass(ClassCodeVisitorSupport.java:56)
	at com.microsoft.gradle.compile.CompletionVisitor.lambda$visitModule$1(CompletionVisitor.java:100)
	at java.base/java.lang.Iterable.forEach(Unknown Source)
	at com.microsoft.gradle.compile.CompletionVisitor.visitModule(CompletionVisitor.java:99)
	at com.microsoft.gradle.compile.CompletionVisitor.visitSourceUnit(CompletionVisitor.java:92)
	at com.microsoft.gradle.compile.CompletionVisitor.lambda$visitCompilationUnit$0(CompletionVisitor.java:81)
	at java.base/java.util.Iterator.forEachRemaining(Unknown Source)
	at com.microsoft.gradle.compile.CompletionVisitor.visitCompilationUnit(CompletionVisitor.java:81)
	at com.microsoft.gradle.GradleServices.completion(GradleServices.java:269)
	... 16 more
```
**Does the bug still exist if you disable all other extensions?**
Yes

**Additional context**
still fails with the normal `Gradle For Java` version (failing with both pre-release & release versions on vscode extension store)

--- a/gradle-language-server/src/main/java/com/microsoft/gradle/compile/CompletionVisitor.java
+++ b/gradle-language-server/src/main/java/com/microsoft/gradle/compile/CompletionVisitor.java
@@ -104,13 +104,14 @@ public class CompletionVisitor extends ClassCodeVisitorSupport {
 	@Override
 	public void visitMethodCallExpression(MethodCallExpression node) {
 		this.methodCalls.get(this.currentUri).add(node);
-		if (node.getMethodAsString().equals("dependencies")) {
+		String methodName = node.getMethodAsString();
+		if ("dependencies".equals(methodName)) {
 			this.dependencies.get(this.currentUri).addAll(getDependencies(node));
-		} else if (node.getMethodAsString().equals("plugins")) {
+		} else if ("plugins".equals(methodName)) {
 			// match plugins { id: ${id} }
 			List<String> plugins = getPluginFromPlugins(node);
 			this.plugins.get(this.currentUri).addAll(plugins);
-		} else if (node.getMethodAsString().equals("apply")) {
+		} else if ("apply".equals(methodName)) {
 			// match apply plugins: '${id}'
 			String plugin = getPluginFromApply(node);
 			if (plugin != null) {
@@ -203,7 +204,7 @@ public class CompletionVisitor extends ClassCodeVisitorSupport {
 		if (argument instanceof ArgumentListExpression) {
 			List<Expression> expressions = ((ArgumentListExpression) argument).getExpressions();
 			for (Expression expression : expressions) {
-				if (expression instanceof ConstantExpression && node.getMethodAsString().equals("id")) {
+				if (expression instanceof ConstantExpression && "id".equals(node.getMethodAsString())) {
 					results.add(expression.getText());
 				} else if (expression instanceof ClosureExpression) {
 					Statement code = ((ClosureExpression) expression).getCode();

--- a/gradle-language-server/src/main/java/com/microsoft/gradle/compile/DocumentSymbolVisitor.java
+++ b/gradle-language-server/src/main/java/com/microsoft/gradle/compile/DocumentSymbolVisitor.java
@@ -131,7 +131,7 @@ public class DocumentSymbolVisitor {
 				builder.append(objectText);
 				builder.append(".");
 			}
-			builder.append(expression.getMethodAsString());
+			builder.append(expression.getMethod().getText());
 			Expression arguments = expression.getArguments();
 			if (arguments instanceof ArgumentListExpression) {
 				List<Expression> expressions = ((ArgumentListExpression) arguments).getExpressions();
@@ -149,7 +149,7 @@ public class DocumentSymbolVisitor {
 			StringBuilder builder = new StringBuilder();
 			builder.append(getSymbolName((PropertyExpression) objectExpression));
 			builder.append(".");
-			builder.append(expression.getMethodAsString());
+			builder.append(expression.getMethod().getText());
 			return builder.toString();
 		}
 		return null;
@@ -205,23 +205,21 @@ public class DocumentSymbolVisitor {
 
 	private List<DocumentSymbol> getDependencies(MethodCallExpression expression) {
 		Expression argument = expression.getArguments();
-		String name = expression.getMethodAsString();
+		String name = expression.getMethod().getText();
 		if ("dependencies".equals(name)) {
 			return getDependencies((ArgumentListExpression) argument);
 		}
 		List<DocumentSymbol> results = new ArrayList<>();
-		if (name != null) { // avoid NPE in case of dynamic method calls
-			DocumentSymbol symbol = new DocumentSymbol();
-			symbol.setName(name);
-			String detail = getDetail(expression);
-			if (detail != null) {
-				symbol.setDetail(detail);
-			}
-			symbol.setKind(SymbolKind.Constant);
-			symbol.setRange(LSPUtils.toRange(expression));
-			symbol.setSelectionRange(LSPUtils.toRange(expression));
-			results.add(symbol);
+		DocumentSymbol symbol = new DocumentSymbol();
+		symbol.setName(name);
+		String detail = getDetail(expression);
+		if (detail != null) {
+			symbol.setDetail(detail);
 		}
+		symbol.setKind(SymbolKind.Constant);
+		symbol.setRange(LSPUtils.toRange(expression));
+		symbol.setSelectionRange(LSPUtils.toRange(expression));
+		results.add(symbol);
 		return results;
 	}
 

--- a/gradle-language-server/src/main/java/com/microsoft/gradle/compile/DocumentSymbolVisitor.java
+++ b/gradle-language-server/src/main/java/com/microsoft/gradle/compile/DocumentSymbolVisitor.java
@@ -205,21 +205,23 @@ public class DocumentSymbolVisitor {
 
 	private List<DocumentSymbol> getDependencies(MethodCallExpression expression) {
 		Expression argument = expression.getArguments();
-		if ("dependencies".equals(expression.getMethodAsString())) {
+		String name = expression.getMethodAsString();
+		if ("dependencies".equals(name)) {
 			return getDependencies((ArgumentListExpression) argument);
 		}
 		List<DocumentSymbol> results = new ArrayList<>();
-		DocumentSymbol symbol = new DocumentSymbol();
-		String name = expression.getMethodAsString();
-		symbol.setName(name);
-		String detail = getDetail(expression);
-		if (detail != null) {
-			symbol.setDetail(detail);
+		if (name != null) { // avoid NPE in case of dynamic method calls
+			DocumentSymbol symbol = new DocumentSymbol();
+			symbol.setName(name);
+			String detail = getDetail(expression);
+			if (detail != null) {
+				symbol.setDetail(detail);
+			}
+			symbol.setKind(SymbolKind.Constant);
+			symbol.setRange(LSPUtils.toRange(expression));
+			symbol.setSelectionRange(LSPUtils.toRange(expression));
+			results.add(symbol);
 		}
-		symbol.setKind(SymbolKind.Constant);
-		symbol.setRange(LSPUtils.toRange(expression));
-		symbol.setSelectionRange(LSPUtils.toRange(expression));
-		results.add(symbol);
 		return results;
 	}
 

--- a/gradle-language-server/src/main/java/com/microsoft/gradle/compile/DocumentSymbolVisitor.java
+++ b/gradle-language-server/src/main/java/com/microsoft/gradle/compile/DocumentSymbolVisitor.java
@@ -113,7 +113,7 @@ public class DocumentSymbolVisitor {
 		}
 		symbol.setSelectionRange(LSPUtils.toRange(expression));
 		symbol.setRange(LSPUtils.toRange(expression));
-		if (expression.getMethodAsString().equals("dependencies")) {
+		if ("dependencies".equals(expression.getMethodAsString())) {
 			List<DocumentSymbol> dependencySymbols = getDependencies(expression);
 			symbol.setChildren(dependencySymbols);
 			this.dependencies.get(currentUri).addAll(dependencySymbols);
@@ -205,7 +205,7 @@ public class DocumentSymbolVisitor {
 
 	private List<DocumentSymbol> getDependencies(MethodCallExpression expression) {
 		Expression argument = expression.getArguments();
-		if (expression.getMethodAsString().equals("dependencies")) {
+		if ("dependencies".equals(expression.getMethodAsString())) {
 			return getDependencies((ArgumentListExpression) argument);
 		}
 		List<DocumentSymbol> results = new ArrayList<>();

--- a/gradle-language-server/src/main/java/com/microsoft/gradle/handlers/CompletionHandler.java
+++ b/gradle-language-server/src/main/java/com/microsoft/gradle/handlers/CompletionHandler.java
@@ -49,6 +49,9 @@ public class CompletionHandler {
 			results.addAll(getCompletionItemsFromExtClosures(resolver, projectPath, resultSet));
 		} else {
 			String methodName = containingCall.getMethodAsString();
+			if (methodName == null) {
+				return Collections.emptyList();
+			}
 			List<CompletionItem> re = getCompletionItemsFromExtClosures(resolver, projectPath, methodName, resultSet);
 			results.addAll(re);
 			List<String> delegates = GradleDelegate.getDelegateMap().get(methodName);

